src/webviews/extension-side/dataviewer/kernelDataViewerDependencyImplementation.ts

```diff
@@ -0,0 +1,107 @@
+import { parseSemVer } from '../../../platform/common/utils';
+import { pandasMinimumVersionSupportedByVariableViewer } from './constants';
+import { BaseDataViewerDependencyImplementation } from './baseDataViewerDependencyImplementation';
+
+export const kernelGetPandasVersion =
+    'import pandas as _VSCODE_pandas;print(_VSCODE_pandas.__version__);del _VSCODE_pandas';
+
+function kernelPackaging(kernel: IKernel): '%conda' | '%pip' {
+    const envType = kernel.kernelConnectionMetadata.interpreter?.envType;
+    const isConda = envType === EnvironmentType.Conda;
+    // From https://ipython.readthedocs.io/en/stable/interactive/magics.html#magic-pip (%conda is here as well).
+    return isConda ? '%conda' : '%pip';
+}
+
+type IKernelWithSession = IKernel & { session: IKernelConnectionSession };
+
+// TypeScript will narrow the type to PythonEnvironment in any block guarded by a call to isPythonEnvironment
+function kernelHasSession(kernel: IKernel): kernel is IKernelWithSession {
+    return Boolean(kernel.session);
+}
+
+/**
+ * Uses the Kernel to manage the dependencies of a Data Viewer.
+ */
+export class KernelDataViewerDependencyImplementation extends BaseDataViewerDependencyImplementation {
+    protected async execute(command: string, kernel: IKernelWithSession): Promise<(string | undefined)[]> {
+        const outputs = await executeSilently(kernel.session, command);
+        const error = outputs.find((item) => item.output_type === 'error');
+        if (error) {
+            traceWarning(DataScience.failedToGetVersionOfPandas(), error.message);
+        }
+        return outputs.map((item) => item.text?.toString());
+    }
+
+    protected async getVersion(kernel: IKernelWithSession): Promise<SemVer | undefined> {
+        try {
+            const outputs = await this.execute(kernelGetPandasVersion, kernel);
+            return outputs.map((text) => (text ? parseSemVer(text.toString()) : undefined)).find((item) => item);
+        } catch (e) {
+            traceWarning(DataScience.failedToGetVersionOfPandas(), e.message);
+            return;
+        }
+    }
+
+    private async installMissingDependencies(kernel: IKernelWithSession): Promise<void> {
+        sendTelemetryEvent(Telemetry.PythonModuleInstall, undefined, {
+            action: 'displayed',
+            moduleName: ProductNames.get(Product.pandas)!
+        });
+
+        const command = `${kernelPackaging(kernel)} install pandas`;
+
+        if (await this.promptInstall()) {
+            try {
+                await this.execute(command, kernel);
+                sendTelemetryEvent(Telemetry.UserInstalledPandas);
+            } catch (e) {
+                sendTelemetryEvent(Telemetry.UserInstalledPandas, undefined, undefined, e);
+                throw new Error(DataScience.failedToInstallPandas());
+            }
+        } else {
+            sendTelemetryEvent(Telemetry.UserDidNotInstallPandas);
```

Comment:
Sending the telemetry after installing and when failing, or when not required to install is still duplicated and feel this can be moved into the base class.